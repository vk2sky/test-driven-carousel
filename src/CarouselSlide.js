import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const DefaultImg = styled.img`
  object-fit: cover; 
  width: 100%;
  height: ${props =>
    typeof props.imgHeight === "number"
      ? `${props.imgHeight}px`
      : props.imgHeight};
`;

const CarouselSlide = (
  {
    Img,
    imgUrl,
    imgHeight,
    description,
    attribution,
    ...rest
  }) =>
  <figure {...rest} >
    <Img src={imgUrl} imgHeight={imgHeight}/>

    <figcaption>
      <strong>{description}</strong> {attribution}
    </figcaption>
  </figure>;

CarouselSlide.propTypes = {
  Img: PropTypes.any,
  imgHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  imgUrl: PropTypes.string.isRequired,
  description: PropTypes.node.isRequired,
  attribution: PropTypes.node,
};

CarouselSlide.defaultProps = {
  Img: DefaultImg,
  imgHeight: 500,
};

export default CarouselSlide;