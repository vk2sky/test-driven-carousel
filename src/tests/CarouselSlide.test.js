import React from "react";
import { shallow, mount } from "enzyme";
import styled from "styled-components";

import CarouselSlide from "../CarouselSlide";

describe("CarouselSlide", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <CarouselSlide
        imgUrl="https://example.com/default.jpg"
        description="Default test image"
      />
    );
  });

  it("renders a <figure>", () => {
    expect(wrapper.type()).toBe("figure");
  });

  it("renders an <img> and a <figcaption> as children", () => {
    expect(wrapper.childAt(0).type()).toBe(CarouselSlide.defaultProps.Img);
    expect(wrapper.childAt(1).type()).toBe("figcaption");
  });

  it("passes `imgUrl` through to props.Img", () => {
    const imgUrl = "https://example.com/image.png";
    wrapper.setProps({ imgUrl });
    const img = wrapper.find(CarouselSlide.defaultProps.Img);
    expect(img.prop("src")).toBe(imgUrl);
  });

  it("uses `description` and `attribution` as the <figcaption>", () => {
    const description = "A jaw-droppingly spectacular image";
    const attribution = "Trevor Burnham";
    wrapper.setProps({ description, attribution });

    expect(wrapper.find("figcaption").text())
      .toBe(`${description} ${attribution}`);

    expect(wrapper.find("figcaption strong").text())
      .toBe(description);
  });

  it("passes other props through to the <figcaption>", () => {
    const style = {};
    const onClick = () => {};
    const className = "my-carousel-slide";
    wrapper.setProps({ style, onClick, className });

    expect(wrapper.prop("style")).toBe(style);
    expect(wrapper.prop("onClick")).toBe(onClick);
    expect(wrapper.prop("className")).toBe(className);
  });
});

describe("Img", () => {
  let mounted;
  const imgUrl = "https://example.com/default.jpg";

  beforeEach(() => {
    mounted = mount(
      <CarouselSlide.defaultProps.Img
        src={imgUrl}
        imgHeight={500}
      />
    );
  });

  it("renders an <img> with the given src", () => {
    expect(mounted.containsMatchingElement(<img src={imgUrl} />)).toBe(true);
  });

  it("has the expected static styles", () => {
    expect(mounted).toHaveStyleRule("width", "100%");
    expect(mounted).toHaveStyleRule("object-fit", "cover");
  });

  it("uses imgHeight as the height style property", () => {
    expect(mounted).toHaveStyleRule("height", "500px");

    mounted.setProps({ imgHeight: "calc(100vh - 100px)" });
    expect(mounted).toHaveStyleRule("height", "calc(100vh - 100px)");
  });

  it("allows styles to be overridden", () => {
    const TestImg = styled(CarouselSlide.defaultProps.Img)`
      width: auto;
      height: auto;
      object-fit: fill;
    `;

    mounted = mount(
      <CarouselSlide
        Img={TestImg}
        imgUrl={imgUrl}
        description="This prop is required"
      /> );

    expect(mounted.find(TestImg)).toHaveStyleRule("width", "auto");
    expect(mounted.find(TestImg)).toHaveStyleRule("height", "auto");
    expect(mounted.find(TestImg)).toHaveStyleRule("object-fit", "fill");
  });
});
