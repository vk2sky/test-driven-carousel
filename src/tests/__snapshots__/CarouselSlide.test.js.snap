// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselSlide renders correctly 1`] = `
<figure>
  <CarouselSlide__DefaultImg
    imgHeight={500}
    src="https://example.com/default.jpg"
  />
  <figcaption>
    <strong>
      Description
    </strong>
     
    Attribution
  </figcaption>
</figure>
`;

exports[`Img renders correctly 1`] = `
.c0 {
  object-fit: cover;
  width: 100%;
  height: 500px;
}

<img
  className="c0"
  src="https://example.com/default.jpg"
/>
`;
